File: backend/app/schemas/billing.py
from pydantic import BaseModel

from app.models.enums import SubscriptionTier


class CheckoutSessionRequest(BaseModel):
    tier: SubscriptionTier


class CheckoutSessionOut(BaseModel):
    checkout_url: str


class BillingPortalOut(BaseModel):
    portal_url: str


class PlanOut(BaseModel):
    tier: SubscriptionTier
    price: float
    currency: str
    available: bool
    # Sent for the same reason the price is: the landing page had its own
    # copy of these and they went stale the moment the plans were resized.
    credits: int


class SubscriptionStatusOut(BaseModel):
    tier: SubscriptionTier
    # False for a free user, and also for a paid one whose subscription was
    # cancelled but whose paid period hasn't run out yet — in that state there
    # is nothing left to manage, so the UI must not offer to manage it.
    has_subscription: bool
    currency: str
    # Served rather than hardcoded in the UI: the real price lives on the
    # Paystack plan, and a figure duplicated in the React tree is a figure
    # that will eventually disagree with what the card is actually charged.
    plans: list[PlanOut]


class CreditCostsOut(BaseModel):
    """What one of each chargeable action costs.

    Served for the same reason the prices are. The pricing page shows what a
    month of credits buys, and that arithmetic has to use the numbers the
    charging code actually uses — a second copy in the UI would quietly start
    promising a count the product does not honour.
    """

    unlock: int
    tailor: int
    concierge: int


class CreditPackOut(BaseModel):
    code: str
    credits: int
    price: float
    currency: str


class CreditPackRequest(BaseModel):
    pack: str

File: frontend/src/lib/api/billing.ts
import { api } from "../api-client";
import type { CreditCosts, CreditPack, Plan, SubscriptionStatus, SubscriptionTier } from "../types";

export const billingApi = {
  creditPacks: () => api.get<CreditPack[]>("/api/billing/credit-packs"),
  buyCredits: (pack: string) =>
    api.post<{ checkout_url: string }>("/api/billing/buy-credits", { pack }),
  checkoutSession: (tier: Exclude<SubscriptionTier, "free">) =>
    api.post<{ checkout_url: string }>("/api/billing/checkout-session", { tier }),
  /** Paystack's hosted manage page: update the card, or cancel there. */
  portal: () => api.get<{ portal_url: string }>("/api/billing/portal"),
  cancel: () => api.post<void>("/api/billing/cancel"),
  status: () => api.get<SubscriptionStatus>("/api/billing/status"),
  /** Public — the landing page's pricing table, so it can't drift from what
      the card is actually charged. */
  plans: () => api.get<Plan[]>("/api/billing/plans", { skipAuth: true }),
  /** Public, like plans: what a month of credits buys is part of choosing one. */
  creditCosts: () => api.get<CreditCosts>("/api/billing/credit-costs", { skipAuth: true }),
};
